storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3109,10 +3109,1764 @@ buf_block_t *fsp_get_latched_xdes_page(
+      err= flst_validate(header, FSP_HEADER_OFFSET + FSP_FULL_FRAG, &mtr);
+    if (err == DB_SUCCESS)
+      err= flst_validate(header, FSP_HEADER_OFFSET + FSP_SEG_INODES_FULL, &mtr);
+    if (err == DB_SUCCESS)
+      err= flst_validate(header, FSP_HEADER_OFFSET + FSP_SEG_INODES_FREE, &mtr);
+  }
+  mtr.commit();
+  return err;
+}
+
+dberr_t indexDefragmenter::store_child_pages(buf_block_t *block) noexcept
+{
+  const byte *page= block->page.frame;
+  const rec_t *rec= page_rec_get_next_low(page + PAGE_OLD_INFIMUM, false);
+  while (rec != page + PAGE_OLD_SUPREMUM)
+  {
+    ulint len;
+    ulint offset= rec_get_nth_field_offs_old(rec,
+                                             rec_get_n_fields_old(rec) - 1,
+                                             &len);
+    if (len != 4)
+      return DB_CORRUPTION;
+
+    if (offset >= srv_page_size)
+      return DB_CORRUPTION;
+
+    const byte *field= rec + offset;
+    /* m_parent_pages[child_page_no] =
+       1st 32 bit to indicate offset in parent page
+       2nd 32 bit to indicate parent page number */
+    m_parent_pages[mach_read_from_4(field)]=
+      uint64_t(page_offset(field)) << 32 | block->page.id().page_no();
+    rec= page_rec_get_next_low(rec, false);
+  }
+  return DB_SUCCESS;
+}
+
+uint32_t indexDefragmenter::get_level_block(uint16_t level, mtr_t *mtr,
+                                            dberr_t *err) noexcept
+{
+  uint32_t child_page_no= m_index->page;
+  while (1)
+  {
+    buf_block_t *block= fsp_get_latched_page(page_id_t{0, child_page_no},
+                                             mtr, err);
+    if (!block)
+    {
+      child_page_no= FIL_NULL;
+      break;
+    }
+
+    page_t *page= buf_block_get_frame(block);
+    uint16_t cur_level= btr_page_get_level(page);
+    if (cur_level == level)
+      break;
+    const rec_t *rec= page_rec_get_next_low(page + PAGE_OLD_INFIMUM, false);
+    if (rec && rec != page + PAGE_OLD_SUPREMUM)
+    {
+      ulint len;
+      ulint offset=
+        rec_get_nth_field_offs_old(rec, rec_get_n_fields_old(rec) - 1,
+                                   &len);
+      ut_ad(len == 4);
+
+      ut_ad(offset < srv_page_size);
+      child_page_no= mach_read_from_4(rec + offset);
+    }
+    else
+    {
+      child_page_no= FIL_NULL;
+      *err= DB_CORRUPTION;
+      break;
+    }
+    if (cur_level == level + 1)
+      break;
+  }
+  return child_page_no;
+}
+
+dberr_t indexDefragmenter::defragment_level(uint16_t level,
+                                            mtr_t *mtr,                                                          spaceDefragmenter *space_defrag) noexcept
+{
+  dberr_t err= DB_SUCCESS;
+  uint32_t cur_page_no= get_level_block(level, mtr, &err);
+  if (cur_page_no == FIL_NULL)
+  {
+    mtr->commit();
+    return err;
+  }
+
+  buf_block_t *root=
+    buf_pool.page_fix(page_id_t{0, m_index->page}, &err,
+                      buf_pool_t::FIX_WAIT_READ);
+  if (!root)
+  {
+    mtr->commit();
+    return err;
+  }
```

Comment:
Our only caller `defragment()` is already acquiring a root page latch (which we may be releasing in a `mtr->commit()` inside this function). I’d suggest that it ensure that `root` be buffer-fixed outside `mtr->m_memo`, and pass it as a parameter or via a data member to this member function. That would simplify the logic here.